wireapp/wire-android
# fix: not switching to proper conversation when navigating from notification [WPB-4294]

----
#### PR Submission Checklist for internal contributors

- The **PR Title**

- The **PR Description**
----

# What's new in this PR?

### Issues

Tapping on notification does not forward the user to the according conversation when another conversation is open.

### Causes (Optional)

Compose destinations library has two interfaces containing `route` parameter that can't be used interchangeably.
`Route` is used to define destinations and its route is always a template with arguments placeholders, like the native compose navigation uses, for example: `conversation_screen/{conversationId}`. As documentation states: `Route` instances are not suited to navigate to unless they're also `Direction`.
`Direction` also has a `route` parameter, but it's the specific route with filled proper arguments, like `conversation_screen/id123456` as it's used when navigating. So, `route` parameters of both types shouldn't be compared.
Because it couldn't compare it properly, it couldn't find the destination of the same type in the back stack. Together with `launchSingleTop` set to true, when we have already a destination of type `ConverstationScreen` at the top of the back stack, it won't be replaced with a new one, thus when opening another `ConversationScreen` it still keeps the previous one on screen and ignores the new one.

### Solutions

Instead of comparing full routes, extract the base of each route, which is unique for each destination, and then compare only that part to find the specific destination type in the back stack.

### Testing

#### How to Test

Open conversation A, receive notification for conversation B, click on that notification - the app should open the conversation B screen.

----
#### PR Post Submission Checklist for internal contributors (Optional)

----
#### PR Post Merge Checklist for internal contributors

----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.
